tests/test_utils.py

```diff
@@ -68,3 +69,35 @@ def test_hms_to_seconds(seconds: str, expected: int):
     result = hms_to_seconds(seconds)
 
     assert result == expected
+
+
+def test_url_connection_raises_on_invalid_url():
+    with pytest.raises(ValueError, match=r".*invalidhosturl.*"):
+        check_url_connection("foo-invalidhosturl")
+
+
+def test_url_connection_checks_hostname(mocker):
+    socket_library_mock = mocker.patch("bonfire.utils.socket.socket")
+    socket_mock = socket_library_mock.return_value.__enter__.return_value
+    check_url_connection("https://validhost.com")
+    socket_mock.connect.assert_called_with(("validhost.com", 443))
+    socket_mock.settimeout.assert_called_once()
+
+
+def test_url_connection_timeout_handling(mocker):
+    socket_library_mock = mocker.patch("bonfire.utils.socket.socket")
+    socket_mock = socket_library_mock.return_value.__enter__.return_value
+    socket_mock.connect.side_effect = TimeoutError("timed out!")
+
+    with pytest.raises(FatalError, match=r"Unable to connect to.*after.*seconds.*is VPN needed.*"):
+        check_url_connection("https://timingout.com")
+
+
+def test_ip_timeout():
+    with pytest.raises(FatalError, match="Unable to connect to.*after 1 seconds.*is VPN needed.*"):
+        check_url_connection("https://10.255.255.1", timeout=1)
+
+
+def test_url_connection_dns_lookup_fails():
+    with pytest.raises(FatalError, match=r".*DNS lookup failed.*"):
+        check_url_connection("https://baddomain.test")
```

Comment:
I don't think there's a reason not to mock this and use a side effect to raise a `socket.gaierror`, this way we don't need to get creative with domain names if `baddomain.test` ever exists :)

Reply:
Looks like a domain with `.invalid` should never resolve as per the RFC's: https://serverfault.com/a/846523 -- I'll switch to that.

Reply:
Nice ! Thanks